docs/customtut_parsers.rst

```diff
@@ -457,125 +457,202 @@ The following shows our completed parser including documentation.
+    Port 22
+    #AddressFamily any
+    ListenAddress 10.110.0.1
+    Port 22
+    ListenAddress 10.110.1.1
+    #ListenAddress ::
 
-    $ python -m insights.parsers.ssh
+    # The default requires explicit activation of protocol 1
+    #Protocol 2
+    Protocol 1
+    """
 
-If no errors are displayed then ``doctest`` was successful. To run
-``pytest`` on just the ``ssh`` parser execute the following command::
+    def test_sshd_config():
+        sshd_config = SshDConfig(context_wrap(SSHD_CONFIG_INPUT))
+        assert sshd_config is not None
+        assert 'Port' in sshd_config
+        assert 'PORT' in sshd_config
+        assert sshd_config['port'] == ['22', '22']
+        assert 'ListenAddress' in sshd_config
+        assert sshd_config['ListenAddress'] == ['10.110.0.1', '10.110.0.1']
+        assert sshd_config['Protocol'] == ['1']
+        assert 'AddressFamily' not in sshd_config
+        ports = [l for l in sshd_config if l.keyword == 'Port']
+        assert len(ports) == 2
+        assert ports[0].value == '22'
+
+To test the documentation, we can then use ``doctest``:
+
+.. code-block:: python
+    :linenos:
+    :lineno-start: 35
+
+    def test_sshd_documentation():
+        """
+        Here we test the examples in the documentation automatically using
+        doctest.  We set up an environment which is similar to what a
+        rule writer might see - a 'shared' object that stores parser
+        objects by class.  This saves doing this setup in the example code.
+        """
+        env = {
+            'SshDConfig': ssh.SshDConfig,
+            'shared': {
+                ssh.SshDConfig: ssh.SshDConfig(context_wrap(SSHD_CONFIG_INPUT)),
+            }
+        }
+        failed, total = doctest.testmod(ssh, globs=env)
+        assert failed == 0
+
+The environment setup allows us to 'hide' the set-up of the environment that
+normally provided to the rule, which is the context in which the example
+code is written.  This must provide both a name for the parser class, as if
+it had been imported by the rule, and the 'shared' dictionary with a
+reference, by class, to the object created by parsing the sample input.
+
+The ``assert`` line here makes sure that any failures in the examples are
+detected by pytest.  This will also include the testing output from doctest,
+showing where the code failed to evaluate or where the output differed from
+what was given.
+
+Because this code essentially duplicates many of the things previously
+tested explicitly in the ``test_sshd_config`` function, we can remove some
+of those tests and only test the 'corner cases':
+
+.. code-block:: python
+    :linenos:
+    :lineno-start: 51
+
+    def test_sshd_corner_cases():
+        """
+        Here we test any corner cases for behaviour we expect to deal with
+        in the parser but doesn't make a good example.
+        """
+        config = ssh.SshDConfig(context_wrap(SSHD_DOCS_EXAMPLE))
```

Comment:
Works well with `SSHD_CONFIG_INPUT`, if we keep `SSHD_DOCS_EXAMPLE`, it may confuse users

Reply:
`SSHD_DOCS_EXAMPLE` should actually be a separate definition, so I've included that.  Thanks!